com.ibm.streamsx.topology/opt/python/include/splpy_funcop.h

```diff
@@ -131,6 +186,92 @@ class SplpyFuncOp : public SplpyOp {
           SplpyGeneral::callVoidFunction(
               "streamsx.topology.runtime", "setupOperator", tkDir, NULL);
       }
+
+      /**
+       * Support for saving an operator's state to checkpoints, and restoring
+       * the state from checkpoints.
+       */
+      class SplPyFuncOpStateHandler: public SPL::StateHandler {
+      public:
+        // Steals reference to pickledCallable
+        SplPyFuncOpStateHandler(SplpyOp * pyop, PyObject * pickledCallable) : op(pyop), loads(), dumps(), pickledInitialCallable(pickledCallable) {
+          // Load pickle.loads and pickle.dumps
+          SplpyGIL lock;
+          loads = SplpyGeneral::loadFunction("dill", "loads");
+          dumps = SplpyGeneral::loadFunction("dill", "dumps");
+        }
+
+        virtual ~SplPyFuncOpStateHandler() {
+          Py_CLEAR(loads);
+          Py_CLEAR(dumps);
+          Py_CLEAR(pickledInitialCallable);
+        }
+
+        virtual void checkpoint(SPL::Checkpoint & ckpt) {
+          SPLAPPTRC(L_DEBUG, "checkpoint", "python");
+          SPL::blob bytes;
+          {
+            SplpyGIL lock;
+            PyObject * ret = call(dumps, op->callable());
+            if (!ret) {
+              SplpyGeneral::tracePythonError();
+              throw SplpyGeneral::pythonException("dill.dumps");
+            }
+            pySplValueFromPyObject(bytes, ret);
+            Py_DECREF(ret);
+          }
+          ckpt << bytes;
+        }
+
+        virtual void reset(SPL::Checkpoint & ckpt) {
+          SPLAPPTRC(L_DEBUG, "reset", "python");
+          // Restore the callable from an spl blob
+          SPL::blob bytes;
+          ckpt >> bytes;
+          SplpyGIL lock;
+          PyObject * pickle = pySplValueToPyObject(bytes);
+          PyObject * ret = call(loads, pickle);
+          if (!ret) {
+            SplpyGeneral::tracePythonError();
+            throw SplpyGeneral::pythonException("dill.loads");
+          }
+          // discard the old callable, replace with the newly
+          // unpickled one.
+          Py_DECREF(op->callable());
+          op->setCallable(ret); // reference to ret stolen by op
+        }
+
+       virtual void resetToInitialState() {
+         SPLAPPTRC(L_DEBUG, "resetToInitialState", "python");
+         SplpyGIL lock;
+         PyObject * initialCallable = call(loads, pickledInitialCallable);
+         if (!initialCallable) {
+           SplpyGeneral::tracePythonError();
+           throw SplpyGeneral::pythonException("dill.loads");
+         }
+         Py_DECREF(op->callable());
+         op->setCallable(initialCallable);
+       }
+
+      private:
+        // Call a python callable with a single argument
+        // Caller must hold GILState.
+        PyObject * call(PyObject * callable, PyObject * arg) {
```

Comment:
We already have utility methods to call objcets:

https://github.com/IBMStreams/streamsx.topology/blob/master/com.ibm.streamsx.topology/opt/python/include/splpy_general.h#L162

and at line 181

Reply:
The callFunction method in splpy_general log python errors to standard error, with no information about when they happened, in what operator, and from where in our code.  Diagnosing consistent region issues is very tricky, and I think it is important to log any problems in the application trace to preserve information about what was happening when an error occurred.